### PR TITLE
[PaxosLog] Sqlite implementation without namespaces

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
This is the basic psqlite implementation, without namespacing which will be added in a subsequent PR.

**Implementation Description (bullets)**:
Relatively straightforward, although there is some magic with how the result sets are interpreted when there are no results

**Testing (What was existing testing like?  What have you done to improve it?)**:
Unit tests using in memory implementation

**Concerns (what feedback would you like?)**:
I think it's fine

**Where should we start reviewing?**:
Tests?

**Priority (whenever / two weeks / yesterday)**:
Now

### CODE REVIEW

File: leader-election-impl/src/main/java/com/palantir/paxos/SqliteConnections.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.function.Supplier;
+
+import com.google.common.base.Suppliers;
+
+/**
+ * This class is responsible for creating Sqlite connections to an instance.
+ * There should be one instance per timelock.
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/paxos/SqliteConnections.java:28]: Not now, and we'll change this: should we *enforce* this?

File: leader-election-impl/src/test/java/com/palantir/paxos/SqlitePaxosStateLogTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+import com.google.common.base.Suppliers;
+
+public class SqlitePaxosStateLogTest {
+    private PaxosStateLog<PaxosValue> stateLog;
+
+    @Before
+    public void setup() {
+        Supplier<Connection> connectionSupplier = Suppliers.memoize(SqliteConnections.createDatabaseForTest()::get);
+        stateLog =  SqlitePaxosStateLog.createInitialized(connectionSupplier);
```
Comment (jeremyk-91) [inline leader-election-impl/src/test/java/com/palantir/paxos/SqlitePaxosStateLogTest.java:37]:
```suggestion
        stateLog = SqlitePaxosStateLog.createInitialized(connectionSupplier);
```

File: leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    public static <V extends Persistable & Versionable> PaxosStateLog<V> createInitialized(Supplier<Connection> conn) {
+        SqlitePaxosStateLog<V> log = new SqlitePaxosStateLog<>(conn);
+        log.initialize();
+        return log;
+    }
+
+    private void initialize() {
+        executeVoid("CREATE TABLE IF NOT EXISTS paxosLog (seq BIGINT, val BLOB, CONSTRAINT pk_dual PRIMARY KEY (seq))");
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java:45]: will be fixed in next PR: the constraint is funky

File: leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            preparedStatement.setBytes(2, round.persistToBytes());
+            preparedStatement.execute();
+        } catch (SQLException e) {
+            throw Throwables.rewrapAndThrowUncheckedException(e);
+        }
+    }
+
+    @Override
+    public byte[] readRound(long seq) {
+        return executeStatement(String.format("SELECT val FROM paxosLog WHERE seq = %s;", seq))
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java:63]:
```suggestion
        return executeStatement(String.format("SELECT val FROM paxosLog WHERE seq = %s", seq))
```

File: leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            preparedStatement.setBytes(2, round.persistToBytes());
+            preparedStatement.execute();
+        } catch (SQLException e) {
+            throw Throwables.rewrapAndThrowUncheckedException(e);
+        }
+    }
+
+    @Override
+    public byte[] readRound(long seq) {
+        return executeStatement(String.format("SELECT val FROM paxosLog WHERE seq = %s", seq))
```
Comment (j-baker) [inline leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java:63]: please can we use jooq here? Direct Java sql is kind of shitty...

Comment (j-baker) [reply leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java:63]: or JDBI - JDBI's pretty great.

Comment (j-baker) [reply leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java:63]: Explicitly, something like this:

```java
public interface UserDao {
    @SqlUpdate("CREATE TABLE user (id INTEGER PRIMARY KEY, name VARCHAR)")
    void createTable();

    @SqlUpdate("INSERT INTO user(id, name) VALUES (?, ?)")
    void insertPositional(int id, String name);

    @SqlUpdate("INSERT INTO user(id, name) VALUES (:id, :name)")
    void insertNamed(@Bind("id") int id, @Bind("name") String name);

    @SqlUpdate("INSERT INTO user(id, name) VALUES (:id, :name)")
    void insertBean(@BindBean User user);

    @SqlQuery("SELECT * FROM user ORDER BY name")
    @RegisterBeanMapper(User.class)
    List<User> listUsers();
}
```

File: leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    private void executeVoid(String statement) {
+        try {
+            connectionSupplier.get().prepareStatement(statement).execute();
+        } catch (SQLException e) {
+            throw Throwables.rewrapAndThrowUncheckedException(e);
+        }
+    }
+
+    private Optional<ResultSet> executeStatement(String statement) {
```
Comment (j-baker) [inline leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java:95]: Connection is closeable :)

Comment (j-baker) [reply leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java:95]: JDBI handles that for you :)

File: leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    @Override
+    public byte[] readRound(long seq) {
+        return executeStatement(String.format("SELECT val FROM paxosLog WHERE seq = %s", seq))
+                .map(SqlitePaxosStateLog::getByteArrayUnchecked)
+                .orElse(null);
+    }
+
+    @Override
+    public long getLeastLogEntry() {
+        return executeStatement("SELECT MIN(seq) FROM paxosLog")
```
Comment (jkozlowski) [inline leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java:70]: Do we not need an index?

File: leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.function.Supplier;
+
+import com.google.common.io.ByteStreams;
+import com.palantir.common.base.Throwables;
+import com.palantir.common.persist.Persistable;
+
+public final class SqlitePaxosStateLog<V extends Persistable & Versionable> implements PaxosStateLog<V> {
+    private final Supplier<Connection> connectionSupplier;
+
+    private SqlitePaxosStateLog(Supplier<Connection> connectionSupplier) {
```
Comment (jkozlowski) [inline leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLog.java:34]: Who closes the connection?

**Pull request merged**